plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -422,11 +473,339 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+	// Choose the one(s) with lowest sum of priorities.
+	minNumPods := math.MaxInt32
+	lowestSumPriorityNodes := []*nodeScore{}
+	for _, nodeScore := range lowestHighPriorityNodes {
+		if nodeScore.sumPriorities == minSumPriorities {
+			lowestSumPriorityNodes = append(lowestSumPriorityNodes, nodeScore)
+			if nodeScore.numPods < minNumPods {
+				minNumPods = nodeScore.numPods
+			}
+		}
+	}
+	if len(lowestSumPriorityNodes) == 1 {
+		return lowestSumPriorityNodes[0].nodeName
+	}
+	// There are still more than one node with minimum highest priority victim and
+	// lowest sum of victim priorities. Find the anyone with minimum number of victims.
+	for _, nodeScore := range lowestSumPriorityNodes {
+		if nodeScore.numPods == minNumPods {
+			return nodeScore.nodeName
+		}
+	}
+	glog.Errorf("We should never reach here!")
+	return ""
+}
+
+// selectNodesForPreemption finds all the nodes with possible victims for
+// preemption in parallel.
+func selectNodesForPreemption(pod *v1.Pod,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+	nodes []*v1.Node,
+	potentialNodeNames map[string]bool,
+	predicates map[string]algorithm.FitPredicate,
+	metadataProducer algorithm.PredicateMetadataProducer,
+	extenders []algorithm.SchedulerExtender,
+) (map[string][]*v1.Pod, error) {
+
+	nodeNameToPods := map[string][]*v1.Pod{}
+	var resultLock sync.Mutex
+
+	// We can use the same metadata producer for all nodes.
+	meta := metadataProducer(pod, nodeNameToInfo)
+	checkNode := func(i int) {
+		nodeName := nodes[i].Name
+		// if potentialNodeNames is not nil, try the node only if it exists in potentialNodeNames.
+		// potentialNodeNames may be nil only in tests.
+		if potentialNodeNames != nil {
+			if _, ok := potentialNodeNames[nodeName]; !ok {
+				return
+			}
+		}
+		var metaCopy algorithm.PredicateMetadata
+		if meta != nil {
+			metaCopy = meta.ShallowCopy()
+		}
+		pods, fits := selectVictimsOnNode(pod, metaCopy, nodeNameToInfo[nodeName], predicates, extenders)
+		if fits {
+			resultLock.Lock()
+			nodeNameToPods[nodeName] = pods
+			resultLock.Unlock()
+		}
+	}
+	workqueue.Parallelize(16, len(nodes), checkNode)
+
+	// If there are any extenders, run them and filter the list of candidate nodes.
+	if len(nodeNameToPods) > 0 && len(extenders) > 0 {
+		filteredNodeNameToPods := map[string][]*v1.Pod{}
+		// Remove the victims from the corresponding nodeInfo and send nodes to the
+		// extenders for filtering.
+		for nodeName, victims := range nodeNameToPods {
+			originalNodeInfo := nodeNameToInfo[nodeName]
+			nodeInfoCopy := nodeNameToInfo[nodeName].Clone()
+			for _, victim := range victims {
+				nodeInfoCopy.RemovePod(victim)
+			}
+			nodeNameToInfo[nodeName] = nodeInfoCopy
+			filteredNodes := nodes
+			nodePassedExtenders := true
+			for _, extender := range extenders {
+				var err error
+				var failedNodesMap map[string]string
+				filteredNodes, failedNodesMap, err = extender.Filter(pod, filteredNodes, nodeNameToInfo)
```

Comment:
I'm concerned about its performance. This means we are calling extender (for all nodes) for every single candidate for preemption (which may potentially be all nodes). So this is effectively a quadratic algorithm.

Given that we are going to introduce an extender by default soon (PVC), I'm not sure if this is feasible approach.
We need to be really careful about this part.

@davidopp

Reply:
The short-term alternative is to always assume we can't preempt for extender feasibility failures, and then invert the logic in nodesWherePreemptionMightHelp()  so it assumes failures it doesn't know about are non-preemptible. Then you don't have to call out to the extender (i.e. go back to the version of this part of the PR before the extender call-out was added). See my comment here: https://github.com/kubernetes/kubernetes/pull/50949#discussion_r135399169
I think that should work OK.

But longer-term we DO want to preempt for local PV feasibility failures (even if we don't care about other extenders) so we need to think about how to do it properly.

One thing I am not 100% sure about is if the local PV folks decided to use the regular extender mechanism for local PV, or a new mechanism that is more efficient (requires fewer roundtrips). The benchmarking results looked good so you are probably right that they are going to use the regular extender mechanism, but we should verify.

Reply:
I think I'm too tired to understand your comment. I will get to it tomorrow morning.

One thing I'm not aware - what is "the new extender mechanism"? I'm not aware of anything like that.

Reply:
This afternoon @bsalamat and I talked with @msau42 and she said she is probably going to integrate local PV scheduling into the default scheduler directly (as a library, so the code can easily be reused by custom schedulers) rather than implementing it as an extender. Obviously we should still be concerned about extender performance but this would mean it is still only used in cluster that explicitly opt in by running an extender. (Also, none of the local PV stuff is going into 1.8 in any event.)

@bsalamat filed #51656 to find an approach that allows extender to perform preemption and presumably also improve the performance beyond the revised approach he is using in this PR now.

Reply:
Yeah I'm going with the library approach even though extender performance looked fine.  The main issue is the challenges and ramifications of exposing controller-manager as a public extender endpoint in an HA configuration.